README.md
Confluent Platform
==================

Description
-----------

Apache Kafka, an open source technology created and maintained by the founders
of Confluent, acts as a real-time, fault tolerant, highly scalable messaging
system. It is widely adopted for use cases ranging from collecting user
activity data, logs, application metrics, stock ticker data, and device
instrumentation. Its key strength is its ability to make high volume data
available as a real-time stream for consumption in systems with very different
requirements—from batch systems like Hadoop, to real-time systems that require
low-latency access, to stream processing engines that transform the data
streams as they arrive.

This infrastructure lets you build around a single central nervous system
transmitting messages to all the different systems and applications within your
company. Learn more on <http://confluent.io>.

This cookbook focuses on deploying Confluent Platform elements on your clusters
via Chef on *systemd* managed distributions. At the moment, this includes
**Kafka**, **Schema Registry** and **Kafka Rest**.

Requirements
------------

### Cookbooks and gems

Declared in [metadata.rb](metadata.rb) and in [Gemfile](Gemfile).

### Platforms

A *systemd* managed distribution:
- RHEL Family 7, tested on Centos

Note: it should work fine on Debian 8 but the official docker image does not
allow systemd to work easily, so it could not be tested.

Usage
-----

### Easy Setup

Default recipe does nothing. Each service **Kafka**, **Schema Registry** or
**Kafka Rest** will be installed by calling respectively recipe
[install-kafka](recipes/install-kafka.rb),
[install-registry](recipes/install-registry.rb) and
[install-rest](recipes/install-rest.rb).

By default, this cookbook installs *openjdk* from the official repositories
*(openjdk 8 on centos 7)* in **services** recipe, just before
launching the service. You can deactivate this behavior by setting
`node['confluent-platform']['java']` to `""`, or choose your package by setting
the package name in `node['confluent-platform']['java'][node[:platform]]`.

### Search

The recommended way to use this cookbook is through the creation of a different
role per cluster, that is a role for **Kafka**, **Schema Registry** and
**Kafka Rest**. This enables the search by role feature, allowing a simple
service discovery.

See [roles](test/integration/roles) for some examples and *Cluster Search*
documentation for more information.

### Test

This cookbook is fully tested through the installation of the full platform
in docker hosts. This uses kitchen, docker and some monkey-patching.

If you run `kitchen list`, you will see 7 suites:
- dnsdock-centos-7
- zookeeper-centos-7
- kafka-01-centos-7
- kafka-02-centos-7
- kafka-03-centos-7
- registry-01-centos-7
- rest-01-centos-7

Each corresponds to a different node in the cluster.

For more information, see [.kitchen.yml](.kitchen.yml) and [test](test)
directory.

### Local cluster

Of course, the cluster you install by running `kitchen converge` is fully
working so you can use it as a local cluster to test your development (like a
new Kafka client). Moreover, compared to a single node cluster usually
installed on workstations, you can detected partition/timing/fault-tolerance
issues you could not because of the simplicity of a single-node system.

You can access it by adding the dnsdock used in the cluster as your main DNS
resolver: add
`docker inspect --format '{{.NetworkSettings.IPAddress}}' dnsdock-kafka`
in `/etc/resolv.conf`.

Then to produce some messages:

    kafka-console-producer.sh \
      --broker-list kafka-kitchen-01.kitchen:9092 \
      --topic my_topic

And to read them:

    kafka-console-consumer.sh \
      --zookeeper zookeeper-kafka.kitchen/kafka-kitchen \
      --topic my_topic \
      --from-beginning

Or you can use Rest API with http://rest-kitchen-01.kitchen:8082 and full
Schema Registry support, located at http://registry-kitchen-01.kitchen:8081.

Attributes
----------

Configuration is done by overriding default attributes. All configuration keys
have a default defined in [attributes/default.rb](attributes/default.rb).
Please read it to have a comprehensive view of what and how you can configure
this cookbook behavior.

Recipes
-------

### default

Does nothing.

### repository

Configure confluent repository.

### install-*service*

Install and fully configure a given *service* by running *repository* and its
4 dedicated recipes: *package*, *user*, *config* and *service*, in that order.

### *service*-package

Install given *service* from confluent repository.

### *service*-user

Create given *service* system user and group.

### *service*-config

Generate *service* configuration. May search for dependencies (like Zookeeper
or other nodes of the same cluster) with the help of cluster-search cookbook.

### *service*-service

Install systemd unit for the given *service*, then enable and start it.

Note: install *java* package (OpenJDK 8 on centos 7) by default, can be
disabled by setting `node['confluent-platform']['java']` to "". A platform
specific configuration for the package to install is also possible.

Resources/Providers
-------------------

None.

Changelog
---------

Available in [CHANGELOG](CHANGELOG).

Contributing
------------

Please read carefully [CONTRIBUTING.md](CONTRIBUTING.md) before making a merge
request.

License and Author
------------------

- Author:: Samuel Bernard (<samuel.bernard@s4m.io>)

```text
Copyright:: 2015, Sam4Mobile

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG
Changelog
=========

1.2.0
-----

Main:
- Clarify and fix JVM options for services
- Use to_hash instead of dup to work on node values
- Improve readibility of default system user names

Fixes:
- Fix and clean the creation of Kafka work directories
- Fix zookeeper.connect chroot path

Test:
- Rationalize docker provision to limit images
- Fix typo in roles/rest-kitchen.json name
- Wait 15s after registry start to strengthen tests

Packaging:
- Reorganize README:
  + Move changelog from README to CHANGELOG
  + Move contribution guide to CONTRIBUTING.md
  + Reorder README, fix Gemfile missing
- Add Apache 2 license file
- Add missing chefignore
- Fix long lines in rest and registry templates

1.1.0
-----

- Cleaning, use only dependencies from supermarket

1.0.1
-----

- Set java-1.8.0-openjdk-headless as default java package

1.0.0
-----

- Initial version with Centos 7 support


attributes/default.rb
#
# Copyright (c) 2015 Sam4Mobile
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# confluent version and general cookbook attributes
default['confluent-platform']['version']        = '1.0'
default['confluent-platform']['scala_version']  = '2.11.5'
default['confluent-platform']['java']['centos'] = 'java-1.8.0-openjdk-headless'

# Cluster search configuration
# To understand the following attributes, look at 'cluster-search' README

# Zookeeper cluster
default['confluent-platform']['zookeeper']['role']  = 'zookeeper-cluster'
default['confluent-platform']['zookeeper']['hosts'] = []
default['confluent-platform']['zookeeper']['size']  = 3

# Kafka cluster
default['confluent-platform']['kafka']['role']  = 'kafka-cluster'
default['confluent-platform']['kafka']['hosts'] = []
default['confluent-platform']['kafka']['size']  = 3

# Schema Registry cluster
default['confluent-platform']['registry']['role']  = 'schema-registry-cluster'
default['confluent-platform']['registry']['hosts'] = []
default['confluent-platform']['registry']['size']  = 3

# Kafka Rest cluster
default['confluent-platform']['rest']['role']  = 'kafka-rest-cluster'
default['confluent-platform']['rest']['hosts'] = []
default['confluent-platform']['rest']['size']  = 3

# Kafka configuration
# Always use a chroot in Zookeeper
default['confluent-platform']['kafka']['zk_chroot'] =
  "/#{node['confluent-platform']['kafka']['role']}"

default['confluent-platform']['kafka']['user'] = 'kafka'
default['confluent-platform']['kafka']['auto_restart'] = 'true'

# Kafka configuration, default provided by Kafka project
default['confluent-platform']['kafka']['config']      = {
  'broker.id' => 0,
  'port' => 9092,
  'num.network.threads' => 3,
  'num.io.threads' => 8,
  'socket.send.buffer.bytes' => 102400,
  'socket.receive.buffer.bytes' => 102400,
  'socket.request.max.bytes' => 104857600,
  'log.dirs' => '/var/lib/kafka',
  'num.partitions' => 1,
  'num.recovery.threads.per.data.dir' => 1,
  'log.retention.hours' => 168,
  'log.segment.bytes' => 1073741824,
  'log.retention.check.interval.ms' => 300000,
  'log.cleaner.enable' => false,
  'zookeeper.connect' => 'localhost:2181',
  'zookeeper.connection.timeout.ms' => 6000
}

# Kafka JVM configuration
default['confluent-platform']['kafka']['jvm_opts'] = {
  '-Xms4g' => nil,
  '-Xmx4g' => nil,
  '-XX:+UseG1GC' => nil,
  '-XX:MaxGCPauseMillis' => 20,
  '-XX:InitiatingHeapOccupancyPercent' => 35
}

# Kafka JMX configuration
default['confluent-platform']['kafka']['jmx_opts'] = {
  '-Dcom.sun.management.jmxremote' => nil,
  '-Dcom.sun.management.jmxremote.authenticate' => false,
  '-Dcom.sun.management.jmxremote.ssl' => false,
  '-Dcom.sun.management.jmxremote.port' => 8090,
  '-Djava.rmi.server.hostname' => node['fqdn']
}

# Kafka log4j configuration
default['confluent-platform']['kafka']['log4j'] = {
  'kafka.logs.dir' => '/var/log/kafka',
  'log4j.rootLogger' => 'INFO, stdout ',
  'log4j.appender.stdout' => 'org.apache.log4j.ConsoleAppender',
  'log4j.appender.stdout.layout' => 'org.apache.log4j.PatternLayout',
  'log4j.appender.stdout.layout.ConversionPattern' => '[%d] %p %m (%c)%n',
  'log4j.appender.kafkaAppender' =>
    'org.apache.log4j.DailyRollingFileAppender',
  'log4j.appender.kafkaAppender.DatePattern' => "'.'yyyy-MM-dd-HH",
  'log4j.appender.kafkaAppender.File' => '${kafka.logs.dir}/server.log',
  'log4j.appender.kafkaAppender.layout' => 'org.apache.log4j.PatternLayout',
  'log4j.appender.kafkaAppender.layout.ConversionPattern' =>
    '[%d] %p %m (%c)%n',
  'log4j.appender.stateChangeAppender' =>
    'org.apache.log4j.DailyRollingFileAppender',
  'log4j.appender.stateChangeAppender.DatePattern' =>
    "'.'yyyy-MM-dd-HH",
  'log4j.appender.stateChangeAppender.File' =>
    '${kafka.logs.dir}/state-change.log',
  'log4j.appender.stateChangeAppender.layout' =>
    'org.apache.log4j.PatternLayout',
  'log4j.appender.stateChangeAppender.layout.ConversionPattern' =>
    '[%d] %p %m (%c)%n',
  'log4j.appender.requestAppender' =>
    'org.apache.log4j.DailyRollingFileAppender',
  'log4j.appender.requestAppender.DatePattern' => "'.'yyyy-MM-dd-HH",
  'log4j.appender.requestAppender.File' =>
    '${kafka.logs.dir}/kafka-request.log',
  'log4j.appender.requestAppender.layout' => 'org.apache.log4j.PatternLayout',
  'log4j.appender.requestAppender.layout.ConversionPattern' =>
    '[%d] %p %m (%c)%n',
  'log4j.appender.cleanerAppender' =>
    'org.apache.log4j.DailyRollingFileAppender',
  'log4j.appender.cleanerAppender.DatePattern' => "'.'yyyy-MM-dd-HH",
  'log4j.appender.cleanerAppender.File' => '${kafka.logs.dir}/log-cleaner.log',
  'log4j.appender.cleanerAppender.layout' => 'org.apache.log4j.PatternLayout',
  'log4j.appender.cleanerAppender.layout.ConversionPattern' =>
    '[%d] %p %m (%c)%n',
  'log4j.appender.controllerAppender' =>
    'org.apache.log4j.DailyRollingFileAppender',
  'log4j.appender.controllerAppender.DatePattern' => "'.'yyyy-MM-dd-HH",
  'log4j.appender.controllerAppender.File' =>
    '${kafka.logs.dir}/controller.log',
  'log4j.appender.controllerAppender.layout' =>
    'org.apache.log4j.PatternLayout',
  'log4j.appender.controllerAppender.layout.ConversionPattern' =>
    '[%d] %p %m (%c)%n',
  'log4j.logger.kafka' => 'INFO, kafkaAppender',
  'log4j.logger.kafka.network.RequestChannel$' => 'WARN, requestAppender',
  'log4j.additivity.kafka.network.RequestChannel$' => 'false',
  'log4j.logger.kafka.request.logger' => 'WARN, requestAppender',
  'log4j.additivity.kafka.request.logger' => 'false',
  'log4j.logger.kafka.controller' => 'TRACE, controllerAppender',
  'log4j.additivity.kafka.controller' => 'false',
  'log4j.logger.kafka.log.LogCleaner' => 'INFO, cleanerAppender',
  'log4j.additivity.kafka.log.LogCleaner' => 'false',
  'log4j.logger.state.change.logger' => 'TRACE, stateChangeAppender',
  'log4j.additivity.state.change.logger' => 'false'
}


# Schema Registry configuration
default['confluent-platform']['registry']['user'] = 'schema-registry'
default['confluent-platform']['registry']['auto_restart'] = 'true'
default['confluent-platform']['registry']['config'] = {
  'port' => '8081',
  'kafkastore.connection.url' => 'localhost:2181',
  'kafkastore.topic' => '_schemas',
  'debug' => 'false'
}

default['confluent-platform']['registry']['log4j'] = {
  'log4j.rootLogger' => 'INFO, stdout',
  'log4j.appender.stdout' => 'org.apache.log4j.ConsoleAppender',
  'log4j.appender.stdout.layout' => 'org.apache.log4j.PatternLayout',
  'log4j.appender.stdout.layout.ConversionPattern' => '[%d] %p %m (%c:%L)%n',
  'log4j.logger.kafka' => 'ERROR, stdout',
  'log4j.logger.org.apache.zookeeper' => 'ERROR, stdout',
  'log4j.logger.org.apache.kafka' => 'ERROR, stdout',
  'log4j.logger.org.I0Itec.zkclient' => 'ERROR, stdout',
  'log4j.additivity.kafka.server' => 'false',
  'log4j.additivity.kafka.consumer.ZookeeperConsumerConnector' => 'false'
}

# Schema Registry JVM configuration
default['confluent-platform']['registry']['jvm_opts'] = {
  '-Xms1g' => nil,
  '-Xmx1g' => nil,
  '-XX:+UseG1GC' => nil,
  '-XX:MaxGCPauseMillis' => 20,
  '-XX:InitiatingHeapOccupancyPercent' => 35
}

# Schema Registry JMX configuration
default['confluent-platform']['registry']['jmx_opts'] = {
  '-Dcom.sun.management.jmxremote' => nil,
  '-Dcom.sun.management.jmxremote.authenticate' => false,
  '-Dcom.sun.management.jmxremote.ssl' => false,
  '-Dcom.sun.management.jmxremote.port' => 8091,
  '-Djava.rmi.server.hostname' => node['fqdn']
}

# Kafka Rest configuration
default['confluent-platform']['rest']['user'] = 'kafka-rest'
default['confluent-platform']['rest']['auto_restart'] = 'true'
default['confluent-platform']['rest']['config'] = {}
default['confluent-platform']['rest']['log4j'] = {
  'log4j.rootLogger' => 'INFO, stdout',
  'log4j.appender.stdout' => 'org.apache.log4j.ConsoleAppender',
  'log4j.appender.stdout.layout' => 'org.apache.log4j.PatternLayout',
  'log4j.appender.stdout.layout.ConversionPattern' => '[%d] %p %m (%c:%L)%n'
}

# Kafka Rest JVM configuration
default['confluent-platform']['rest']['jvm_opts'] = {
  '-Xms1g' => nil,
  '-Xmx1g' => nil,
  '-XX:+UseG1GC' => nil,
  '-XX:MaxGCPauseMillis' => 20,
  '-XX:InitiatingHeapOccupancyPercent' => 35
}

# Kafka Rest JMX configuration
default['confluent-platform']['rest']['jmx_opts'] = {
  '-Dcom.sun.management.jmxremote' => nil,
  '-Dcom.sun.management.jmxremote.authenticate' => false,
  '-Dcom.sun.management.jmxremote.ssl' => false,
  '-Dcom.sun.management.jmxremote.port' => 8092,
  '-Djava.rmi.server.hostname' => node['fqdn']
}
